From 5c485937cd9d82791aa19621d8bb97042bf60c99 Mon Sep 17 00:00:00 2001
From: yuyue <yuyue@mprc.pku.edu.cn>
Date: Fri, 2 Mar 2012 21:09:47 +0800
Subject: [PATCH 316/641] UniCore64: Implement mcount() to print function
 address

With -pg option, unicore64-linux-gcc will insert mcount() at the
start of functions which don't have attribute no_instrument_function.

Signed-off-by: yuyue <yuyue@mprc.pku.edu.cn>
---
 arch/unicore64/lib/Makefile |    1 +
 arch/unicore64/lib/mcount.S |   14 ++++++++++++++
 2 files changed, 15 insertions(+)
 create mode 100644 arch/unicore64/lib/mcount.S

diff --git a/arch/unicore64/lib/Makefile b/arch/unicore64/lib/Makefile
index 0583feb..fc16a84 100644
--- a/arch/unicore64/lib/Makefile
+++ b/arch/unicore64/lib/Makefile
@@ -1,2 +1,3 @@
 lib-y			:= delay.o debug.o
 lib-y			+= string-simple.o
+lib-y			+= mcount.o
diff --git a/arch/unicore64/lib/mcount.S b/arch/unicore64/lib/mcount.S
new file mode 100644
index 0000000..95b0198
--- /dev/null
+++ b/arch/unicore64/lib/mcount.S
@@ -0,0 +1,14 @@
+#include <linux/linkage.h>
+#include <arch/asm-common.h>
+
+ENTRY(mcount)
+	__push	r0
+	__push	r1
+	ldd     r1, [ip-], #8
+        dsub    r1, r1, #4
+	mov	r0, #0
+	bkpt
+	__pop	r1
+	__pop	r0
+	return
+ENDPROC(mcount)
-- 
1.7.9.5

